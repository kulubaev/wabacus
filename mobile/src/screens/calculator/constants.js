import { StyleSheet } from 'react-native';

export const OPERAND = 'OPERAND';


export const OPERANDS=['0','1','2','3','4','5','6','7','8','9','.'];
export const BINARY_OPERATORS=['+','-','/','^','*'];
export const UNARY_OPERATORS=['!', 'cbrt','sqrt'];
export const CLEAR_OPERATORS=['Escape', 'clr'] ;
export const RESULT_OPERATORS= ['Enter', '='];


const style = StyleSheet.create({
  button:{
    flex: 2,
  }

})

export const buttons = [
  [ 
    { text: '+' },
    { text: '-' }, 
    { text: '*' },
    { text: '/' },
  ],
  [ 
    { text: '7' },
    { text: '8' }, 
    { text: '9' },
    { text: '√', operator: 'sqrt' },
  ],

  [ 
    { text: '4' },
    { text: '5' }, 
    { text: '6' },
    { text: '3√', operator: 'cbrt' },
  ],

  [ 
    { text: '3' },
    { text: '2' }, 
    { text: '0' },
    { text: '!' },
  ],

  [ 
    { text: '1' },
    { text: '.' },
    { text: 'clr', style }, 
  ],


  [ 
    { text: '=', style}, 
  ],
]
