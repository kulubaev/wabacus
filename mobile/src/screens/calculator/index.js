import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { 
  Text,
  TextInput,
  View,
  SafeAreaView,
  StyleSheet

} from 'react-native';

import {
  OPERAND,
  buttons
} from './constants';

import Display from './components/display';
import Pad from './components/pad';

import { connect } from 'react-redux';
import * as actions from './store/actions'; 

import { 
  isClearOperator,
  isResultOperator,
  isBinaryOperator,
  isUnaryOperator,
  isLastBinaryOperator, 
  isLastOperand, 
  lastOperand,
  isOperand,
  operands
} from './utils/operators';


class Calculator extends Component {

  handleUnaryOperator = (operator) => {
    const { infix, Calculate } = this.props;

    if(isLastOperand(infix)) {
      Calculate(operator);
    } 
  }

  handleBinaryOperator = (operator) => {
    const { 
      infix,  
      OverrideLast, 
      UpdateNew,
      Calculate
    } = this.props;
    
    if(isLastBinaryOperator(infix)) {
      OverrideLast(operator);
    } else {
      const opnds = operands(infix);

      if(opnds.length === 2) {
        Calculate(operator);
      }

      if(opnds.length < 2) {
        UpdateNew(operator);
      }
    }
  }

  handleOperand = (operand) => {
    const { infix,  OverrideLast, UpdateNew, editing } = this.props;

    if(isLastOperand(infix)) {

      if(editing){
        const last = infix[infix.length - 1];
        OverrideLast(last + `${operand}`);
      } else {
        OverrideLast(operand);
      }

    }else {
      UpdateNew(operand);
    }
  }

  handleEvent = (value) => {

    switch(true) {

      case isOperand(value):
        this.handleOperand(value);
        break;

      case isBinaryOperator(value):
        this.handleBinaryOperator(value);
        break;

      case isUnaryOperator(value):
        this.handleUnaryOperator(value);
        break;

      case isResultOperator(value):
        this.props.Calculate();
        break;

       case isClearOperator(value):
        this.props.Clear()
        break;

      default:
        break;
    }
  }

  handlePress = (value, e) => {
    e.stopPropagation();
    this.handleEvent(value);
  }

  render() {
    const { infix } = this.props;
    const last = lastOperand(infix);

    return (
      <SafeAreaView style={styles.container}>
        <View style={styles.display}>
          <Display value={last}/>
        </View>
        <View style={styles.keysPad}>
          <Pad  buttons={buttons} onPress={this.handlePress} />
        </View>
        </SafeAreaView>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: '#ddd',

  },
  display: {
    flex: 3,
  },
  keysPad: {
    flex: 9,
  }
});

Calculator.propTypes = {
  infix: PropTypes.array,
  OverrideLast: PropTypes.func,
  UpdateNew: PropTypes.func,
  Calculate: PropTypes.func,
  editing: PropTypes.bool
}

Calculator.defaultProps = {
  infix: [0],
  editing: false
}

const mapToProps = (state) => {
  return {
  editing: state.calculator.editing,
  infix: state.calculator.infix
}};



export default connect(mapToProps, actions)(Calculator);
