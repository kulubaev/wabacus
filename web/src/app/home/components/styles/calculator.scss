.calculator {
  //border: 1px solid #ccc;
  border-radius: 5px;
  width: 400px;
  position: relative;
  background-color: #f8f7f7;
}

.calculator-screen {
  width: 100%;
  font-size: 2rem;
  height: 80px;
  border: none;
  color: #252525;
  background-color: #fff;
  text-align: right;
  padding-right: 20px;
  padding-left: 10px;
  box-sizing: border-box;
  color: #333;
  border: 1px solid #adadad;
}

button {
  height: 45px;
  background-color: #fff;
  border-radius: 3px;
  border: 1px solid #c4c4c4;
  background-color: transparent;
  font-size: 1.5rem;
  color: #333;
	color: #adadad;

  background-image: linear-gradient(to bottom,transparent,transparent 50%,rgba(0,0,0,.04));
  box-shadow: inset 0 0 0 1px rgba(255,255,255,.05), inset 0 1px 0 0 rgba(255,255,255,.45), inset 0 -1px 0 0 rgba(255,255,255,.15), 0 1px 0 0 rgba(255,255,255,.15);
  text-shadow: 0 1px rgba(233,233,233,.9);
}

button:hover {
  background-color: #eaeaea;
  color: #333;
}

.operator {
  color: #333;
}

.all-clear {
  background-image: linear-gradient(to bottom,transparent,transparent 50%,rgba(0,0,0,.04));
  box-shadow: inset 0 0 0 1px rgba(255,255,255,.05), inset 0 1px 0 0 rgba(255,255,255,.45), inset 0 -1px 0 0 rgba(255,255,255,.15), 0 1px 0 0 rgba(255,255,255,.15);
  color: #333;
}

.all-clear:hover {

}

.equal-sign {
  background-color: #efefef;
  background-image: linear-gradient(to bottom,transparent,transparent 50%,rgba(0,0,0,.04));
 
  border-color: #337cac;

  border: 1px solid #c4c4c4;
  color: #fec726;
  height: 100%;
  grid-area: 5 / 5 / 7 / 3;
  color: #333;
}

.calculator-keys {
  display: grid;
  grid-template-columns: repeat(4, 1fr);
  grid-gap: 15px;
  padding: 25px;
}
