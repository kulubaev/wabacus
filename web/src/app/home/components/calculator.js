import React, {Component} from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { 
  isLastBinaryOperator, 
  isLastOperand, 
  lastOperand,
  operands
} from '../utilities/operators';

import * as actions from '../store/actions'; 
import './styles/calculator.scss';

export class Calculator extends Component {

  handleClear = (e) => {
    e.preventDefault();
    this.props.Clear()
  }

  handleResult = (e) => {
    this.props.Calculate();
  }


  handleUnaryOperator = (e) => {
    e.preventDefault();
    const operator = e.target.value;

    const { infix, Calculate } = this.props;

    if(isLastOperand(infix)) {
      Calculate(operator);
    } 
  }


  handleBinaryOperator = (e) => {
    e.preventDefault();
    const operator = e.target.value;

    const { 
      infix,  
      OverrideLast, 
      UpdateNew,
      Calculate
    } = this.props;
    
    if(isLastBinaryOperator(infix)) {
      OverrideLast(operator);
    } else {
      const opnds = operands(infix);

      if(opnds.length === 2) {
        Calculate(operator);
      }

      if(opnds.length < 2) {
        UpdateNew(operator);
      }
    }
  }


  handleOperand = (e) => {
    e.preventDefault();
    const operand = e.target.value;
    const { infix,  OverrideLast, UpdateNew, editing } = this.props;

    if(isLastOperand(infix)) {

      if(editing){
        const last = infix[infix.length - 1];
        OverrideLast(last + `${operand}`);
      } else {
        OverrideLast(operand);
      }

    }else {
      UpdateNew(operand);
    }
  }

  render()  {
    const { infix } = this.props;
    const last = lastOperand(infix);

    return (
      <div className="calculator">
        <input type="text" className="calculator-screen" value={last} disabled />
        <div className="calculator-keys">

          <button onClick={this.handleBinaryOperator} type="button" className="operator" value="+">+</button>
          <button onClick={this.handleBinaryOperator} type="button" className="operator" value="-">-</button>
          <button onClick={this.handleBinaryOperator} type="button" className="operator" value="*">&times;</button>
          <button onClick={this.handleBinaryOperator} type="button" className="operator" value="/">&divide;</button>

          <button onClick={this.handleOperand} type="button" value="7">7</button>
          <button onClick={this.handleOperand} type="button" value="8">8</button>
          <button onClick={this.handleOperand} type="button" value="9">9</button>
          <button onClick={this.handleUnaryOperator} type="button" className="operator" value="$">&radic;</button>

          <button onClick={this.handleOperand} type="button" value="4">4</button>
          <button onClick={this.handleOperand} type="button" value="5">5</button>
          <button onClick={this.handleOperand} type="button" value="6">6</button>
          <button onClick={this.handleUnaryOperator} type="button" className="operator" value="#">&sup3;&radic;</button>

          <button onClick={this.handleOperand} type="button" value="3">3</button>
          <button onClick={this.handleOperand} type="button" value="2">2</button>
          <button onClick={this.handleOperand} type="button" value="0">0</button>

          <button onClick={this.handleUnaryOperator} type="button" className="operator" value="!">!</button>

          <button onClick={this.handleOperand} type="button" value="1">1</button>

          <button onClick={this.handleBinaryOperator} type="button" className="operator" value="^">x<sup>y</sup></button>

          <button onClick={this.handleOperand} type="button" value=".">.</button>

          <button onClick={this.handleClear} type="button" className="all-clear" value="">clr</button>
          <button onClick={this.handleResult}type="button" className="equal-sign" value="=">=</button>
        </div>
      </div>
    );
  }
}

Calculator.propTypes = {
  infix: PropTypes.array,
  OverrideLast: PropTypes.func,
  UpdateNew: PropTypes.func,
  Calculate: PropTypes.func,
  editing: PropTypes.bool
}

Calculator.defaultProps = {
  infix: [0],
  editing: false
}

const mapToProps = (state) => {
  return {
  editing: state.home.editing,
  infix: state.home.infix
}};

export default connect(mapToProps, actions)(Calculator);

