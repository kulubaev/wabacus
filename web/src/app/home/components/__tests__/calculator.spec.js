import React from 'react';
import renderer from 'react-test-renderer';
import { Calculator } from '../calculator';

import { mount } from 'enzyme';



describe('calculator component', () => {

  /**
   * UI related test
   * utilizes feetures like 
   * regressing testing agains snapshots
   */

  describe('ui verification', () =>  {

    it ('should not regress and render ui correctly', () => {

      const shot = renderer
        .create(<Calculator/>)
        .toJSON();

      expect(shot).toMatchSnapshot();

    });
  });

  describe('behavior  verification', () => {

    /**
     * intenally stack data structure 
     * has been used to mimick 
     * well known postfix mnemonics
     */


    let props = {};
    let sut = null;

    beforeEach(() => {

      props = {
        OverrideLast: jest.fn((v) => { 
          const infix = sut.prop('infix');
          infix[infix.length -1] = v;
          sut.setProps({editing:true, infix});
        }),

        UpdateNew: jest.fn((v) => { 
          const infix = sut.prop('infix');
          infix.push(v);
          sut.setProps({editing:true, infix});
        }),
        Calculate: jest.fn()
      };

      sut = mount(<Calculator {...props}/>);
     
    });

    it ('should be able to perform unary as expected ', () => {


      /*
       * testing:
       * expression: 24!
       */

      sut.find('button[value="2"]').simulate('click');

      expect(props.OverrideLast.mock.calls.length).toEqual(1);
      expect(props.OverrideLast.mock.calls[0]).toEqual(['2']);


      expect(sut.prop('infix')).toEqual(['2']);
      expect(sut.prop('editing')).toEqual(true);


      sut.find('button[value="4"]').simulate('click');

      expect(props.OverrideLast.mock.calls.length).toEqual(2);
      expect(props.OverrideLast.mock.calls[1]).toEqual(['24']);

      expect(props.Calculate.mock.calls.length).toEqual(0);
      expect(props.UpdateNew.mock.calls.length).toEqual(0);

      expect(sut.prop('infix')).toEqual(['24']);
      expect(sut.prop('editing')).toEqual(true);


      sut.find('button.operator[value="!"]').simulate('click');
      expect(props.Calculate.mock.calls.length).toEqual(1);
    })



    it ('should be able to perform binary operator as expected ', () => {

      /*
       * testing:
       * expression: 3 + 5
       */

      sut.find('button[value="3"]').simulate('click');

      expect(props.OverrideLast.mock.calls.length).toEqual(1);
      expect(props.OverrideLast.mock.calls[0]).toEqual(['3']);

      expect(props.Calculate.mock.calls.length).toEqual(0);
      expect(props.UpdateNew.mock.calls.length).toEqual(0);

      expect(sut.prop('infix')).toEqual(['3']);


      sut.find('button.operator[value="+"]').simulate('click');

      expect(props.UpdateNew.mock.calls.length).toEqual(1);
      expect(props.UpdateNew.mock.calls[0]).toEqual(['+']);

      expect(props.Calculate.mock.calls.length).toEqual(0);
      expect(props.OverrideLast.mock.calls.length).toEqual(1);


      expect(sut.prop('infix')).toEqual(['3', '+']);

      sut.find('button[value="5"]').simulate('click');

      expect(props.UpdateNew.mock.calls.length).toEqual(2);
      expect(props.UpdateNew.mock.calls[1]).toEqual(['5']);
      expect(props.Calculate.mock.calls.length).toEqual(0);


      expect(sut.prop('infix')).toEqual(['3', '+', '5']);


      sut.find('button[value="="]').simulate('click');
      expect(props.Calculate.mock.calls.length).toEqual(1);

      expect(props.UpdateNew.mock.calls.length).toEqual(2);
      expect(props.OverrideLast.mock.calls.length).toEqual(1);

    });

  });



});


