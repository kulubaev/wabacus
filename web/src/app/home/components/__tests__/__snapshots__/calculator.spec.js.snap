// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculator component ui verification should not regress and render ui correctly 1`] = `
<div
  className="calculator"
>
  <input
    className="calculator-screen"
    disabled={true}
    type="text"
    value={0}
  />
  <div
    className="calculator-keys"
  >
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="+"
    >
      +
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="-"
    >
      -
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="*"
    >
      ×
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="/"
    >
      ÷
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="7"
    >
      7
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="8"
    >
      8
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="9"
    >
      9
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="$"
    >
      √
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="4"
    >
      4
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="5"
    >
      5
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="6"
    >
      6
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="#"
    >
      ³√
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="3"
    >
      3
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="2"
    >
      2
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="0"
    >
      0
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="!"
    >
      !
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="1"
    >
      1
    </button>
    <button
      className="operator"
      onClick={[Function]}
      type="button"
      value="^"
    >
      x
      <sup>
        y
      </sup>
    </button>
    <button
      onClick={[Function]}
      type="button"
      value="."
    >
      .
    </button>
    <button
      className="all-clear"
      onClick={[Function]}
      type="button"
      value=""
    >
      clr
    </button>
    <button
      className="equal-sign"
      onClick={[Function]}
      type="button"
      value="="
    >
      =
    </button>
  </div>
</div>
`;
